doc/config-generic-update-rollback/SONiC_Generic_Config_Update_and_Rollback_Design.md

```diff
@@ -0,0 +1,867 @@
+- Config rollback should be with minimum disruption to the device e.g. if reverting ACL updates DHCP should not be affected i.e. minimum-disruption rollback
+- User should be able to preview the configuration difference before going ahead and committing the configuration changes
+- In case of errors, the system should just report an error and the user should take care of it
+- Only one session globally can update device config at a time i.e. no concurrent updates to configurations
+
+### 1.1.2 Configuration and Management Requirements
+- All commands argument to generated using Python-click to provide help menus and other out-of-the box features
+- Only root user must be allowed to execute the commands
+- Non-root users can execute commands with dry-run option
+- Each command must provide the following sub options:
+  - "dry-run" Perform a dry-run of the command showing exactly what will be executed on the device, without executing it
+  - "verbose" Provide additional information on the steps executed
+
+### 1.1.3 Scalability Requirements
+N/A
+
+### 1.1.4 Warm Boot Requirements
+N/A
+
+## 1.2 Design Overview
+
+<img src="files/sonic-design.png" alt="sonic-design" width="1200"/>
+
+### 1.2.1 Basic Approach
+SONiC ConfigDB contents can be retrieved in a JSON file format. Modifying JSON file should follow some rules in order to make it straightforward for the users. Fortunately there is already a formal way of defining JSON config updates. It is called JsonPatch, and is formally defined in [RFC 6902 JSON Patch](https://tools.ietf.org/html/rfc6902).
+
+On top of ConfigDBConnector we are going to implement [RFC 6902 JSON Patch](https://tools.ietf.org/html/rfc6902). This API we will call `apply-patch`. On top of that API, we will implement the `rollback` functionality. It will simply starts by getting the diff (patch) between the checkpoint and the current running config, then it will call the API `apply-patch` to update that patch.
+
+The [JsonPatch](https://pypi.org/project/jsonpatch/) python is an open source library that already implements the [RFC 6902 JSON Patch](https://tools.ietf.org/html/rfc6902). We can leverage this library to verify patch config, generate a diff between checkpoint and current running config, verify apply-patch and rollback work as expected by simulating the final output of the update and comparing with the observed output.
+
+**Example:**
+Assume running-config to be:
+```
+{
+  "DEVICE_NEIGHBOR": {
+    "Ethernet8": {
+      "name": "Servers1",
+      "port": "eth0"
+    },
+    "Ethernet96": {
+      "name": "Servers23",
+      "port": "eth0"
+    },
+  },
+  "DHCP_SERVERS": {
+    "1.1.1.1": {},
+    "2.2.2.2": {},
+    "3.3.3.3": {}
+  }
+}
+```
+
+and the ask to:
+- replace *port* under *Ethernet8* with *eth1*
+- add 4.4.4.4 to DHCP_SERVERS
+- remove 2.2.2.2 from DHCP_SERVERS
+
+The steps would be:
+
+1) Take a checkpoint of the config
```

Comment:
Take the content of config DB here and store it?

Reply:
Sure, will update.